pages/apim/3.x/event-native/apim-event-native-api-management-evolution.adoc

```diff
@@ -0,0 +1,322 @@
+
+NOTE: If no configuration is provided, a default configuration is set to default to 30 000ms timeout.
+
+Timeout can now be triggered on two places in the chain, as follows:
+
+* The flow can be interrupted between the beginning of the request and the end of response API flow. In this case, a platform response flow will be executed.
+* The flow can be interrupted during the platform response flow, because the overall request time is too big, causing a 504 response and getting the platform response flow interrupted.
+
+Two properties are available to address this:
+* `http.requestTimeout` - the duration used to configure the timeout of the request.
+* `http.requestTimeoutGraceDelay` - an additional time used to give the platform response flow a chance to execute.
+
+The timeout value is calculated from those two properties:
+* `Timeout = Max(http.requestTimeoutGraceDelay, http.requestTimeout - apiElapsedTime)`
+* With `apiElapsedTime = System.currentTimeMillis() - request().timestamp()`.
+
+==== Examples
+
+NOTE: In the following examples we assume that there is no timeout defined for the backend in the API's endpoint configuration. In real life, those timeout values should be shorter than `http.requestTimeout`, and should interrupt the flow at invoker level.
+
+We will use `http.requestTimeout=2000ms` and `http.requestTimeoutGraceDelay=30ms`.
+
+The example below shows timelines indicating when a timeout should occur depending on the duration of the API flow and the response platform flows:
+
+image:{% link /images/apim/3.x/event-native/event-native-api-management-timeout.png %}[]
+
+
+
+
+
+
+
+== Plan selection
+
+
+=== Common behavior
+
+The plan selection workflow parses all the published plans in the following order: JWT, Oauth2, ApiKey, Keyless.
+
+The parsed plan is selected for execution if all the following conditions are met:
+* The request contains a token corresponding to this plan type (api-key header, authorization header).
+* The plan condition rule is either not set or set but not truthy.
+* There is an active subscription matching the incoming request.
+
+=== V3 mode
+
+In V3 mode, Oauth2 plan *is selected* even if the incoming request does not match a subscription.
+
+No JWT token introspection is done during Oauth2 plan selection.
+
+If there are multiple Oauth2 plans, that would lead to the selection of an irrelevant one.
+
+=== Event-native mode
+
+In event-native execution mode, Oauth2 plan *is not selected* if the incoming request does not match a subscription.
+
+During the Oauth2 plan selection, a token introspection is released in order to retrieve the `client_id`, which allows searching for a subscription.
+
+Due to concerns about performance, a cache system is available to avoid doing the same token introspection multiple times.
+
+If possible, we encourage using selection rules if there are multiple Oauth2 plans, in order to avoid any unnecessary token introspection.
```

Comment:
is it still accurate @jhaeyaert ?

Reply:
Yes except for the keyless plan where the policy has been changed to avoid its activation in case of any security token has been detected in the request by one of the previous plan.
It means that, if an api has multiple plans as follow:

- JWT plan1
- JWT plan2
- Oauth2 plan1
- Oauth2 plan2
- Oauth2 plan2
- Apikey
- Kelyess

And the incoming request contains a token or an apikey that does not match any plan then the keyless plan will not be activated and the user will receive a generic 401 without any details.

Reply:
Me too.

Reply:
All addressed (I hope)